spec/mongoid/contextual/mongo/documents_loader_spec.rb

```diff
@@ -0,0 +1,135 @@
+# frozen_string_literal: true
+
+require "spec_helper"
+
+describe Mongoid::Contextual::Mongo::DocumentsLoader do
+  let(:view) do
+    double('view').tap do |view|
+      allow(view).to receive(:map)
+    end
+  end
+
+  let(:klass) do
+    double
+  end
+
+  let(:criteria) do
+    double('criteria').tap do |criteria|
+      allow(criteria).to receive(:inclusions).and_return([])
+    end
+  end
+
+  let(:executor) do
+    described_class.executor
+  end
+
+  let(:subject) do
+    described_class.new(view, klass, criteria, executor: executor)
+  end
+
+  context 'state management' do
+    let(:executor) do
+      # Such executor will never execute a task, so it guarantees that
+      # our task will stay in its initial state.
+      Concurrent::ThreadPoolExecutor.new(
+        min_threads: 0,
+        max_threads: 0,
+      )
+    end
+
+    describe '#initialize' do
+      it 'initializes in pending state' do
+        expect(subject.pending?).to be_truthy
+        expect(subject.started?).to be_falsey
+      end
+    end
+
+    describe '#unschedule' do
+      it 'changes state' do
+        subject.unschedule
+        expect(subject.pending?).to be_falsey
+        expect(subject.started?).to be_falsey
+      end
+    end
+
+    describe '#execute' do
+      it 'does not change state' do
+        prev_started = subject.started?
```

Comment:
shouldn't you assert what the state is?

Reply:
Initial state is tested above, in `#initialize` specs. I am not sure if we should re-test it here.

Reply:
I personally would assert initial states.

Reply:
If we add an assertion for the initial state here, and then for some reason setting initial state in the constructor is broken, we  will that the test `Mongoid::Contextual::Mongo::DocumentsLoader#execute does not change state` fails, even though the `#execute` method is fine.
The initial state set by the `#initialize` method which is tested above.